README.md
# Sensirion UART SFA3X Arduino Library

This is the Sensirion SFA3X library for Arduino using the
modules UART interface.

[<center><img src="images/SFA3x.png" width="500px"></center>](https://www.sensirion.com/en/environmental-sensors/evaluation-kit-sek-sfa30/)

Click [here](https://www.sensirion.com/my-sfa-ek/) to learn more about the SFA30
sensor and the SFA30 Evaluation Kit Board.

# Installation

To install, download the latest release as .zip file and add it to your
[Arduino IDE](http://www.arduino.cc/en/main/software) via

	Sketch => Include Library => Add .ZIP Library...

Don't forget to **install the dependencies** listed below the same way via `Add
.ZIP Library`

Note: Installation via the Arduino Library Manager is coming soon.

# Dependencies

* [Sensirion Core](https://github.com/Sensirion/arduino-core)

# Quick Start

1. Connect the SFA3X Sensor to one of your Arduino board's
   UART buses. We recommend using `Serial2`. Check the pinout of your Arduino
   board to find the correct pins. The pinout of the SFA3X
   Sensor board can be found in the data sheet.

	* **VDD** of the SEK-SFA3X to the **3.3V** of your Arduino board (5V is also possible)
	* **GND** of the SEK-SFA3X to the **GND** of your Arduino board
	* **RX** of the SEK-SFA3X to the **UART2-TX** of your Arduino board
	* **TX** of the SEK-SFA3X to the **UART2-RX** of your Arduino board

2. Open the `exampleUsage` sample project within the Arduino IDE

		File => Examples => Sensirion UART Sfa3x => exampleUsage

3. Depending on your Arduino board you may need to adapt the `#define
   SENSOR_SERIAL_INTERFACE` to choose the correct serial interface.

4. Click the `Upload` button in the Arduino IDE or

		Sketch => Upload

5. When the upload process has finished, open the `Serial Monitor` or `Serial
   Plotter` via the `Tools` menu to observe the measurement values. Note that
   the `Baud Rate` in the corresponding window has to be set to `115200 baud`.

# Contributing

**Contributions are welcome!**

We develop and test this driver using our company internal tools (version
control, continuous integration, code review etc.) and automatically
synchronize the master branch with GitHub. But this doesn't mean that we don't
respond to issues or don't accept pull requests on GitHub. In fact, you're very
welcome to open issues or create pull requests :)

This Sensirion library uses
[`clang-format`](https://releases.llvm.org/download.html) to standardize the
formatting of all our `.cpp` and `.h` files. Make sure your contributions are
formatted accordingly:

The `-i` flag will apply the format changes to the files listed.

```bash
clang-format -i src/*.cpp src/*.h
```

Note that differences from this formatting will result in a failed build until
they are fixed.

# License

See [LICENSE](LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [unreleased]

### Fixed
- Link to SFA3x Eval Kit overview page

## [0.1.0] - 2021-02-05

Initial release

[Unreleased]: https://github.com/sensirion/arduino-uart-sfa3x/compare/0.1.0...HEAD
[0.1.0]: https://github.com/Sensirion/arduino-uart-sfa3x/releases/tag/0.1.0


src/SensirionUartSfa3x.h
/*
 * THIS FILE IS AUTOMATICALLY GENERATED AND MUST NOT BE EDITED MANUALLY!
 *
 * SHDLC-Generator: 0.8.2
 * Yaml Version: 0.1.0
 * Template Version: 0.3.0
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SENSIRIONUARTSFA3X_H
#define SENSIRIONUARTSFA3X_H

#include <SensirionCore.h>

class SensirionUartSfa3x {

  public:
    SensirionUartSfa3x();
    /**
     * begin() - Initializes the SensirionUartSfa3x class.
     *
     * @param serial Arduino stream object to be communicated with.
     *
     */
    void begin(Stream& serial);

    /**
     * startContinuousMeasurement() - Default output format 0.
     *
     * @note This command is only available in idle mode.
     *
     * @return 0 on success, an error code otherwise
     */
    uint16_t startContinuousMeasurement(void);

    /**
     * stopMeasurement() - Leaves the measurement mode and returns to idle mode.
     *
     * @note This command is only available in measurement mode.
     *
     * @return 0 on success, an error code otherwise
     */
    uint16_t stopMeasurement(void);

    /**
     * readMeasuredValuesOutputFormat2() - Read measured values integer.
     *
     * @param hcho Formaldehyde concentration in ppb as a 16-bit signed integer.
     * Positive values only, negative values are coerced to zero. Scale factor 5
     * ppb^(-1).
     *
     * @param relativeHumidity Relative humidity in %. Values between 0 and 100
     * only. With internal correction for temperature offset (if configured).
     * Scale factor 100 %^(-1).
     *
     * @param temperature Temperature in °C. With internal correction for
     * temperature offset (if configured). Scale factor 200 °C^(-1).
     *
     * @return 0 on success, an error code otherwise
     */
    uint16_t readMeasuredValuesOutputFormat2(int16_t& hcho,
                                             int16_t& relativeHumidity,
                                             int16_t& temperature);

    /**
     * getDeviceMarking() - Gets the device marking from the device.
     *
     * @param deviceMarking String containing the device marking.
     *
     * @return 0 on success, an error code otherwise
     */
    uint16_t getDeviceMarking(uint8_t deviceMarking[],
                              uint8_t deviceMarkingSize);

    /**
     * deviceReset() - Executes a reset on the device.
     *
     * @note The device will reply before executing the reset.
     *
     * @return 0 on success, an error code otherwise
     */
    uint16_t deviceReset(void);

  private:
    Stream* _serial = nullptr;
};

#endif /* SENSIRIONUARTSFA3X_H */


src/SensirionShdlcSfa3x.h
/*
 * THIS FILE IS AUTOMATICALLY GENERATED AND MUST NOT BE EDITED MANUALLY!
 *
 * SHDLC-Generator: 0.8.2
 * Yaml Version: 0.1.0
 * Template Version: 0.2.0
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SENSIRIONSHDLCSFA3X_H
#define SENSIRIONSHDLCSFA3X_H

#include <SensirionCoreArduinoLibrary.h>

class SensirionShdlcSfa3x {

  public:
    SensirionShdlcSfa3x();
    /**
     * begin() - Initializes the SensirionShdlcSfa3x class.
     *
     * @param serial Arduino stream object to be communicated with.
     *
     */
    void begin(Stream& serial);

    /**
     * startContinuousMeasurement() - Default output format 0.
     *
     * @note This command is only available in idle mode.
     *
     * @return 0 on success, an error code otherwise
     */
    uint16_t startContinuousMeasurement(void);

    /**
     * stopMeasurement() - Leaves the measurement mode and returns to idle mode.
     *
     * @note This command is only available in measurement mode.
     *
     * @return 0 on success, an error code otherwise
     */
    uint16_t stopMeasurement(void);

    /**
     * readMeasuredValuesOutputFormat2() - Read measured values integer.
     *
     * @param hcho Formaldehyde concentration in ppb as a 16-bit signed integer.
     * Positive values only, negative values are coerced to zero. Scale factor 5
     * ppb^(-1).
     *
     * @param relativeHumidity Relative humidity in %. Values between 0 and 100
     * only. With internal correction for temperature offset (if configured).
     * Scale factor 100 %^(-1).
     *
     * @param temperature Temperature in °C. With internal correction for
     * temperature offset (if configured). Scale factor 200 °C^(-1).
     *
     * @return 0 on success, an error code otherwise
     */
    uint16_t readMeasuredValuesOutputFormat2(int16_t& hcho,
                                             int16_t& relativeHumidity,
                                             int16_t& temperature);

    /**
     * getDeviceMarking() - Gets the device marking from the device.
     *
     * @param deviceMarking String containing the device marking.
     *
     * @return 0 on success, an error code otherwise
     */
    uint16_t getDeviceMarking(uint8_t deviceMarking[],
                              uint8_t deviceMarkingSize);

    /**
     * deviceReset() - Executes a reset on the device.
     *
     * @note The device will reply before executing the reset.
     *
     * @return 0 on success, an error code otherwise
     */
    uint16_t deviceReset(void);

  private:
    Stream* _serial = nullptr;
};

#endif /* SENSIRIONSHDLCSFA3X_H */


src/SensirionUartSfa3x.cpp
/*
 * THIS FILE IS AUTOMATICALLY GENERATED AND MUST NOT BE EDITED MANUALLY!
 *
 * SHDLC-Generator: 0.8.2
 * Yaml Version: 0.1.0
 * Template Version: 0.3.0
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "SensirionUartSfa3x.h"
#include "Arduino.h"
#include "SensirionCore.h"

#define SFA3X_UART_ADDRESS 0x00

SensirionUartSfa3x::SensirionUartSfa3x() {
}

void SensirionUartSfa3x::begin(Stream& serial) {
    _serial = &serial;
}

uint16_t SensirionUartSfa3x::startContinuousMeasurement() {
    uint16_t error;
    uint8_t buffer[12];
    SensirionShdlcTxFrame txFrame(buffer, 12);
    SensirionShdlcRxFrame rxFrame(buffer, 12);

    error = txFrame.begin(0x00, SFA3X_UART_ADDRESS, 1);
    error |= txFrame.addUInt8(0x00);
    error |= txFrame.finish();
    if (error) {
        return error;
    }

    error = SensirionShdlcCommunication::sendAndReceiveFrame(*_serial, txFrame,
                                                             rxFrame, 10000);

    return error;
}

uint16_t SensirionUartSfa3x::stopMeasurement() {
    uint16_t error;
    uint8_t buffer[12];
    SensirionShdlcTxFrame txFrame(buffer, 12);
    SensirionShdlcRxFrame rxFrame(buffer, 12);

    error = txFrame.begin(0x01, SFA3X_UART_ADDRESS, 0);
    error |= txFrame.finish();
    if (error) {
        return error;
    }

    error = SensirionShdlcCommunication::sendAndReceiveFrame(*_serial, txFrame,
                                                             rxFrame, 10000);

    return error;
}

uint16_t SensirionUartSfa3x::readMeasuredValuesOutputFormat2(
    int16_t& hcho, int16_t& relativeHumidity, int16_t& temperature) {
    uint16_t error;
    uint8_t buffer[24];
    SensirionShdlcTxFrame txFrame(buffer, 24);
    SensirionShdlcRxFrame rxFrame(buffer, 24);

    error = txFrame.begin(0x03, SFA3X_UART_ADDRESS, 1);
    error |= txFrame.addUInt8(0x02);
    error |= txFrame.finish();
    if (error) {
        return error;
    }

    error = SensirionShdlcCommunication::sendAndReceiveFrame(*_serial, txFrame,
                                                             rxFrame, 100000);
    if (error) {
        return error;
    }

    error |= rxFrame.getInt16(hcho);
    error |= rxFrame.getInt16(relativeHumidity);
    error |= rxFrame.getInt16(temperature);
    return error;
}

uint16_t SensirionUartSfa3x::getDeviceMarking(uint8_t deviceMarking[],
                                              uint8_t deviceMarkingSize) {
    uint16_t error;
    uint8_t buffer[522];
    SensirionShdlcTxFrame txFrame(buffer, 522);
    SensirionShdlcRxFrame rxFrame(buffer, 522);

    error = txFrame.begin(0xD0, SFA3X_UART_ADDRESS, 1);
    error |= txFrame.addUInt8(0x06);
    error |= txFrame.finish();
    if (error) {
        return error;
    }

    error = SensirionShdlcCommunication::sendAndReceiveFrame(*_serial, txFrame,
                                                             rxFrame, 10000);
    if (error) {
        return error;
    }

    error |= rxFrame.getBytes(deviceMarking, deviceMarkingSize);
    return error;
}

uint16_t SensirionUartSfa3x::deviceReset() {
    uint16_t error;
    uint8_t buffer[12];
    SensirionShdlcTxFrame txFrame(buffer, 12);
    SensirionShdlcRxFrame rxFrame(buffer, 12);

    error = txFrame.begin(0xD3, SFA3X_UART_ADDRESS, 0);
    error |= txFrame.finish();
    if (error) {
        return error;
    }

    error = SensirionShdlcCommunication::sendAndReceiveFrame(*_serial, txFrame,
                                                             rxFrame, 200000);

    delay(200);
    return error;
}


src/SensirionShdlcSfa3x.cpp
/*
 * THIS FILE IS AUTOMATICALLY GENERATED AND MUST NOT BE EDITED MANUALLY!
 *
 * SHDLC-Generator: 0.8.2
 * Yaml Version: 0.1.0
 * Template Version: 0.2.0
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "SensirionShdlcSfa3x.h"
#include "Arduino.h"
#include "SensirionCoreArduinoLibrary.h"

#define SFA3X_UART_ADDRESS 0x00

SensirionShdlcSfa3x::SensirionShdlcSfa3x() {
}

void SensirionShdlcSfa3x::begin(Stream& serial) {
    _serial = &serial;
}

uint16_t SensirionShdlcSfa3x::startContinuousMeasurement() {
    uint16_t error;
    uint8_t buffer[12];
    SensirionShdlcTxFrame txFrame(buffer, 12);
    SensirionShdlcRxFrame rxFrame(buffer, 12);

    error = txFrame.begin(0x00, SFA3X_UART_ADDRESS, 1);
    error |= txFrame.addUInt8(0x00);
    error |= txFrame.finish();
    if (error) {
        return error;
    }

    error = SensirionShdlcCommunication::sendAndReceiveFrame(*_serial, txFrame,
                                                             rxFrame, 10000);
    if (error) {
        return error;
    }

    return error;
}

uint16_t SensirionShdlcSfa3x::stopMeasurement() {
    uint16_t error;
    uint8_t buffer[12];
    SensirionShdlcTxFrame txFrame(buffer, 12);
    SensirionShdlcRxFrame rxFrame(buffer, 12);

    error = txFrame.begin(0x01, SFA3X_UART_ADDRESS, 0);
    error |= txFrame.finish();
    if (error) {
        return error;
    }

    error = SensirionShdlcCommunication::sendAndReceiveFrame(*_serial, txFrame,
                                                             rxFrame, 10000);
    if (error) {
        return error;
    }

    return error;
}

uint16_t SensirionShdlcSfa3x::readMeasuredValuesOutputFormat2(
    int16_t& hcho, int16_t& relativeHumidity, int16_t& temperature) {
    uint16_t error;
    uint8_t buffer[24];
    SensirionShdlcTxFrame txFrame(buffer, 24);
    SensirionShdlcRxFrame rxFrame(buffer, 24);

    error = txFrame.begin(0x03, SFA3X_UART_ADDRESS, 1);
    error |= txFrame.addUInt8(0x02);
    error |= txFrame.finish();
    if (error) {
        return error;
    }

    error = SensirionShdlcCommunication::sendAndReceiveFrame(*_serial, txFrame,
                                                             rxFrame, 100000);
    if (error) {
        return error;
    }

    error |= rxFrame.getInt16(hcho);
    error |= rxFrame.getInt16(relativeHumidity);
    error |= rxFrame.getInt16(temperature);
    return error;
}

uint16_t SensirionShdlcSfa3x::getDeviceMarking(uint8_t deviceMarking[],
                                               uint8_t deviceMarkingSize) {
    uint16_t error;
    uint8_t buffer[522];
    SensirionShdlcTxFrame txFrame(buffer, 522);
    SensirionShdlcRxFrame rxFrame(buffer, 522);

    error = txFrame.begin(0xD0, SFA3X_UART_ADDRESS, 1);
    error |= txFrame.addUInt8(0x06);
    error |= txFrame.finish();
    if (error) {
        return error;
    }

    error = SensirionShdlcCommunication::sendAndReceiveFrame(*_serial, txFrame,
                                                             rxFrame, 10000);
    if (error) {
        return error;
    }

    error |= rxFrame.getBytes(deviceMarking, deviceMarkingSize);
    return error;
}

uint16_t SensirionShdlcSfa3x::deviceReset() {
    uint16_t error;
    uint8_t buffer[12];
    SensirionShdlcTxFrame txFrame(buffer, 12);
    SensirionShdlcRxFrame rxFrame(buffer, 12);

    error = txFrame.begin(0xD3, SFA3X_UART_ADDRESS, 0);
    error |= txFrame.finish();
    if (error) {
        return error;
    }

    error = SensirionShdlcCommunication::sendAndReceiveFrame(*_serial, txFrame,
                                                             rxFrame, 200000);
    if (error) {
        return error;
    }

    delay(200);
    return error;
}
